tests/unit/db/database_tests.py

```diff
@@ -0,0 +1,480 @@
+        self.assertTrue(
+            all(x in [raw_ddocs[0]['id'], raw_ddocs[1]['id']]
+                for x in self.db.keys()
+                )
+            )
+        self.assertTrue(
+            all(x in [raw_ddocs[0]['doc'], raw_ddocs[1]['doc']]
+                for x in [ddoc1, ddoc2]
+                )
+            )
+
+    def test_retrieve_design_document_list(self):
+        """
+        Test retrieving a list of design document names
+        """
+        map_func = 'function(doc) {\n emit(doc._id, 1); \n}'
+        data = {'_id': '_design/ddoc01','views': {'view01': {"map": map_func}}}
+        self.db.create_document(data)
+        data = {'_id': '_design/ddoc02','views': {'view02': {"map": map_func}}}
+        self.db.create_document(data)
+        ddoc_list = self.db.list_design_documents()
+        self.assertTrue(all(x in ddoc_list for x in self.db.keys()))
+
+    def test_retrieve_design_document(self):
+        """
+        Test retrieve a specific design document
+        """
+        # Get an empty design document object that does not exist remotely
+        local_ddoc = self.db.get_design_document('_design/ddoc01')
+        self.assertEqual(local_ddoc, {'views': {}})
+
+        # Add the design document to the database
+        map_func = 'function(doc) {\n emit(doc._id, 1); \n}'
+        local_ddoc.add_view('view01', map_func)
+        local_ddoc.save()
+
+        # Get the recently created design document that now exists remotely
+        ddoc = self.db.get_design_document('_design/ddoc01')
+        self.assertEqual(ddoc, local_ddoc)
+
+    def test_retrieve_view_results(self):
+        """
+        Test retrieving Result wrapped output from a design document view
+        """
+        map_func = 'function(doc) {\n emit(doc._id, 1); \n}'
+        data = {'_id': '_design/ddoc01','views': {'view01': {"map": map_func}}}
+        self.db.create_document(data)
+        self.populate_db_with_documents()
+
+        # Test with default Result
+        rslt = self.db.get_view_result('_design/ddoc01', 'view01')
+        self.assertIsInstance(rslt, Result)
+        self.assertEqual(rslt[:1], rslt['julia000'])
+
+        #Test with custom Result
+        rslt = self.db.get_view_result(
+            '_design/ddoc01',
+            'view01',
+            descending=True,
+            reduce=False)
+        self.assertIsInstance(rslt, Result)
+        self.assertEqual(rslt[:1], rslt['julia099'])
+
+    def test_retrieve_raw_view_results(self):
+        """
+        Test retrieving raw output from a design document view
+        """
+        map_func = 'function(doc) {\n emit(doc._id, 1); \n}'
+        data = {'_id': '_design/ddoc01','views': {'view01': {"map": map_func}}}
+        self.db.create_document(data)
+        self.populate_db_with_documents()
+
+        raw_rslt = self.db.get_view_raw_result('_design/ddoc01', 'view01')
+        self.assertIsInstance(raw_rslt, dict)
+        self.assertEqual(len(raw_rslt.get('rows')), 100)
```

Comment:
Is there ever a time when this fails because the server hasn't built the view? probably not...